File: reference/python/monitor/monitor.py
"""AARC v1.1.0 executable reference monitor.

The monitor provides hash-chained runtime events, immutable role/objective/policy
anchors, fail-closed tool authorization, separated Actor/Critic/Judge approval
receipts, and deterministic trace verification. It deliberately does not record
hidden chain-of-thought.
"""

from __future__ import annotations

import copy
import hashlib
import json
import sys
import uuid
from datetime import datetime, timezone
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List, Mapping, Optional, Sequence, TextIO


ZERO_HASH = "0" * 64
SCHEMA_VERSION = "1.1.0"
TERMINAL_EVENTS = {"task.stop", "task.fail"}


class TraceVerificationError(ValueError):
    """Raised when a committed trace violates the AARC runtime contract."""


class ToolAuthorizationError(PermissionError):
    """Raised after a denied tool request has been committed to the trace."""


def canonical_json(value: Any) -> str:
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"))


def sha256_hex(value: Any) -> str:
    if not isinstance(value, str):
        value = canonical_json(value)
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def compute_event_hash(event: Mapping[str, Any]) -> str:
    material = dict(event)
    material.pop("event_hash", None)
    return sha256_hex(material)


@dataclass(frozen=True)
class ToolPermit:
    tool_name: str
    request_event_id: str
    permit_event_id: str


class Monitor:
    """Authoritative AARC event committer for one task run."""

    def __init__(
        self,
        *,
        agent_id: str,
        task_id: str,
        objective: str,
        policy: Mapping[str, Any],
        role: str = "actor",
        root_agent_id: Optional[str] = None,
        parent_agent_id: Optional[str] = None,
        run_id: Optional[str] = None,
        allowed_tools: Iterable[str] = (),
        stream: Optional[TextIO] = None,
    ) -> None:
        if not agent_id or not task_id or not objective or not role:
            raise ValueError("agent_id, task_id, objective, and role are required")
        self.agent_id = agent_id
        self.root_agent_id = root_agent_id or agent_id
        self.parent_agent_id = parent_agent_id
        self.task_id = task_id
        self.run_id = run_id or str(uuid.uuid4())
        self.role = role
        self.objective = objective
        self.policy = copy.deepcopy(dict(policy))
        self.allowed_tools = frozenset(allowed_tools)
        self.stream = stream
        self.sequence = 0
        self.events: List[Dict[str, Any]] = []
        self.anchors = {
            "role_hash": sha256_hex(role),
            "objective_hash": sha256_hex(objective),
            "policy_hash": sha256_hex(self.policy),
        }

    def _timestamp(self) -> str:
        return datetime.now(timezone.utc).isoformat()

    def _emit(
        self,
        event_type: str,
        payload: Optional[Mapping[str, Any]] = None,
    ) -> Dict[str, Any]:
        if self.events and self.events[-1]["event_type"] in TERMINAL_EVENTS:
            raise RuntimeError("cannot emit after a terminal event")
        self.sequence += 1
        event: Dict[str, Any] = {
            "event_id": str(uuid.uuid4()),
            "event_type": event_type,
            "timestamp": self._timestamp(),
            "agent_id": self.agent_id,
            "task_id": self.task_id,
            "run_id": self.run_id,
            "sequence": self.sequence,
            "anchors": dict(self.anchors),
            "prev_event_hash": self.events[-1]["event_hash"] if self.events else ZERO_HASH,
            "payload": copy.deepcopy(dict(payload or {})),
        }
        event["event_hash"] = compute_event_hash(event)
        self.events.append(event)
        if self.stream is not None:
            self.stream.write(canonical_json(event) + "\n")
            self.stream.flush()
        return copy.deepcopy(event)

    def task_start(
        self,
        metadata: Optional[Mapping[str, Any]] = None,
    ) -> Dict[str, Any]:
        if self.events:
            raise RuntimeError("task.start must be the first event")
        payload = {
            "schema_version": SCHEMA_VERSION,
            "root_agent_id": self.root_agent_id,
            "parent_agent_id": self.parent_agent_id,
            "role": self.role,
            "objective_digest": self.anchors["objective_hash"],
            "policy_digest": self.anchors["policy_hash"],
        }
        payload.update(dict(metadata or {}))
        return self._emit("task.start", payload)

    def heartbeat(self, note: Optional[str] = None) -> Dict[str, Any]:
        return self._emit("task.heartbeat", {"note": note} if note else {})

    def snapshot_state(
        self,
        *,
        inputs: Mapping[str, Any],
        outputs: Mapping[str, Any],
        tools: Sequence[Mapping[str, Any]] = (),
        confidence: Optional[Mapping[str, Any]] = None,
        assumptions: Sequence[str] = (),
        evidence_refs: Sequence[str] = (),
    ) -> Dict[str, Any]:
        next_sequence = self.sequence + 1
        state = {
            "schema_version": SCHEMA_VERSION,
            "identity": {
                "agent_id": self.agent_id,
                "root_agent_id": self.root_agent_id,
                "parent_agent_id": self.parent_agent_id,
                "role": self.role,
                "role_hash": self.anchors["role_hash"],
            },
            "intent": {
                "objective_hash": self.anchors["objective_hash"],
                "policy_hash": self.anchors["policy_hash"],
            },
            "trace": {
                "run_id": self.run_id,
                "task_id": self.task_id,
                "step_id": f"step-{next_sequence}",
                "sequence": next_sequence,
                "parent_event_hash": (
                    self.events[-1]["event_hash"] if self.events else ZERO_HASH
                ),
            },
            "inputs": copy.deepcopy(dict(inputs)),
            "outputs": copy.deepcopy(dict(outputs)),
            "tools": [copy.deepcopy(dict(item)) for item in tools],
            "confidence": copy.deepcopy(dict(confidence or {})),
            "policy": copy.deepcopy(self.policy),
            "assumptions": list(assumptions),
            "evidence_refs": list(evidence_refs),
        }
        return self._emit("state.snapshot", state)

    def request_tool(
        self,
        tool_name: str,
        arguments: Mapping[str, Any],
    ) -> ToolPermit:
        args_digest = sha256_hex(dict(arguments))
        request = self._emit(
            "tool.requested",
            {"tool_name": tool_name, "arguments_digest": args_digest},
        )
        decision_payload = {
            "tool_name": tool_name,
            "request_event_id": request["event_id"],
            "arguments_digest": args_digest,
            "policy_hash": self.anchors["policy_hash"],
        }
        if tool_name not in self.allowed_tools:
            self._emit("tool.denied", decision_payload)
            raise ToolAuthorizationError(f"tool denied by policy: {tool_name}")
        authorized = self._emit("tool.authorized", decision_payload)
        return ToolPermit(
            tool_name=tool_name,
            request_event_id=request["event_id"],
            permit_event_id=authorized["event_id"],
        )

    def record_tool_result(
        self,
        permit: ToolPermit,
        result: Any,
        *,
        failed: bool = False,
    ) -> Dict[str, Any]:
        payload = {
            "tool_name": permit.tool_name,
            "request_event_id": permit.request_event_id,
            "permit_event_id": permit.permit_event_id,
            "result_digest": sha256_hex(result),
        }
        return self._emit("tool.failed" if failed else "tool.executed", payload)

    def record_approval_decision(
        self,
        *,
        actor_id: str,
        critic_id: str,
        judge_id: str,
        proposal: Any,
        critique: Any,
        decision: str,
        evidence_refs: Sequence[str],
    ) -> Dict[str, Any]:
        if decision not in {"approved", "rejected"}:
            raise ValueError("decision must be approved or rejected")
        if decision == "approved":
            identities = {actor_id, critic_id, judge_id}
            if len(identities) != 3:
                raise ValueError(
                    "positive approval requires pairwise-distinct Actor/Critic/Judge"
                )
            if not evidence_refs:
                raise ValueError("positive approval requires evidence_refs")
        return self._emit(
            "approval.decided",
            {
                "actor_id": actor_id,
                "critic_id": critic_id,
                "judge_id": judge_id,
                "proposal_hash": sha256_hex(proposal),
                "critique_hash": sha256_hex(critique),
                "decision": decision,
                "evidence_refs": list(evidence_refs),
            },
        )

    def remediate(
        self,
        remediation_id: str,
        strategy: str,
        outcome: str,
    ) -> Dict[str, Any]:
        return self._emit(
            "task.remediate",
            {
                "remediation_id": remediation_id,
                "strategy": strategy,
                "outcome": outcome,
            },
        )

    def task_fail(
        self,
        failure_class: str,
        failure_summary: str,
        impact: str,
    ) -> Dict[str, Any]:
        return self._emit(
            "task.fail",
            {
                "failure_class": failure_class,
                "failure_summary": failure_summary,
                "impact": impact,
            },
        )

    def task_stop(self, summary: Optional[str] = None) -> Dict[str, Any]:
        return self._emit("task.stop", {"summary": summary} if summary else {})


def verify_trace(
    events: Sequence[Mapping[str, Any]],
    *,
    allowed_tools: Optional[Iterable[str]] = None,
    require_terminal: bool = True,
) -> None:
    """Verify AARC v1.1.0 structural and semantic invariants."""
    if not events:
        raise TraceVerificationError("empty trace")
    if events[0].get("event_type") != "task.start":
        raise TraceVerificationError("trace must begin with task.start")

    expected_anchors = events[0].get("anchors")
    if not isinstance(expected_anchors, Mapping):
        raise TraceVerificationError("missing genesis anchors")

    permits: Dict[str, Dict[str, Any]] = {}
    denied_requests = set()
    terminal_seen = False
    allowed = frozenset(allowed_tools) if allowed_tools is not None else None
    previous_hash = ZERO_HASH

    for index, raw in enumerate(events, start=1):
        event = dict(raw)
        if terminal_seen:
            raise TraceVerificationError("event appears after terminal event")
        if event.get("sequence") != index:
            raise TraceVerificationError("non-contiguous event sequence")
        if event.get("prev_event_hash") != previous_hash:
            raise TraceVerificationError("broken previous-event hash link")
        if event.get("event_hash") != compute_event_hash(event):
            raise TraceVerificationError("event hash mismatch")
        if event.get("anchors") != expected_anchors:
            raise TraceVerificationError("immutable anchor drift")

        event_type = event.get("event_type")
        payload = event.get("payload")
        if not isinstance(payload, Mapping):
            raise TraceVerificationError("payload must be an object")

        if event_type == "tool.denied":
            request_id = payload.get("request_event_id")
            if request_id:
                denied_requests.add(request_id)

        elif event_type == "tool.authorized":
            request_id = payload.get("request_event_id")
            tool_name = payload.get("tool_name")
            if not request_id or not tool_name:
                raise TraceVerificationError("malformed tool authorization")
            if request_id in denied_requests:
                raise TraceVerificationError("request both denied and authorized")
            if allowed is not None and tool_name not in allowed:
                raise TraceVerificationError("authorization violates verifier policy")
            permits[event["event_id"]] = {
                "request_event_id": request_id,
                "tool_name": tool_name,
            }

        elif event_type in {"tool.executed", "tool.failed"}:
            permit_id = payload.get("permit_event_id")
            request_id = payload.get("request_event_id")
            tool_name = payload.get("tool_name")
            permit = permits.get(permit_id)
            if permit is None:
                raise TraceVerificationError("tool execution without valid permit")
            if (
                permit["request_event_id"] != request_id
                or permit["tool_name"] != tool_name
            ):
                raise TraceVerificationError("tool execution does not match permit")
            if request_id in denied_requests:
                raise TraceVerificationError("denied tool request was executed")
            if allowed is not None and tool_name not in allowed:
                raise TraceVerificationError("executed tool violates verifier policy")

        elif event_type == "approval.decided" and payload.get("decision") == "approved":
            identities = {
                payload.get("actor_id"),
                payload.get("critic_id"),
                payload.get("judge_id"),
            }
            if None in identities or len(identities) != 3:
                raise TraceVerificationError(
                    "self-approval or missing SDV identity"
                )
            if not payload.get("proposal_hash") or not payload.get("critique_hash"):
                raise TraceVerificationError(
                    "approval missing proposal/critique digest"
                )
            if not payload.get("evidence_refs"):
                raise TraceVerificationError("approval missing evidence")

        if event_type in TERMINAL_EVENTS:
            terminal_seen = True

        previous_hash = event["event_hash"]

    if require_terminal and events[-1].get("event_type") not in TERMINAL_EVENTS:
        raise TraceVerificationError("trace is not terminal")


def rehash_trace(
    events: Sequence[Mapping[str, Any]],
) -> List[Dict[str, Any]]:
    """Recompute a chain after intentional mutation in adversarial fixtures."""
    output: List[Dict[str, Any]] = []
    previous_hash = ZERO_HASH
    for raw in events:
        event = copy.deepcopy(dict(raw))
        event["prev_event_hash"] = previous_hash
        event["event_hash"] = compute_event_hash(event)
        output.append(event)
        previous_hash = event["event_hash"]
    return output


def main() -> None:
    monitor = Monitor(
        agent_id="agent-demo",
        task_id="task-demo",
        objective="demonstrate AARC",
        policy={"policy_id": "demo", "constraints": ["read-only"]},
        allowed_tools={"lookup"},
        stream=sys.stdout,
    )
    monitor.task_start()
    monitor.snapshot_state(inputs={"query": "health check"}, outputs={})
    permit = monitor.request_tool("lookup", {"key": "status"})
    monitor.record_tool_result(permit, {"status": "ok"})
    monitor.task_stop("completed")
    verify_trace(monitor.events, allowed_tools={"lookup"})


if __name__ == "__main__":
    main()

File: reference/typescript/monitor/monitor.ts
/* AARC v1.1.0 executable reference monitor for TypeScript. */

import { createHash, randomUUID } from "node:crypto";
import canonicalize from "canonicalize";

type JsonObject = Record<string, unknown>;
type EventRecord = {
  event_id: string;
  event_type: string;
  timestamp: string;
  agent_id: string;
  task_id: string;
  run_id: string;
  sequence: number;
  anchors: {
    role_hash: string;
    objective_hash: string;
    policy_hash: string;
  };
  prev_event_hash: string;
  event_hash: string;
  payload: JsonObject;
};

export const ZERO_HASH = "0".repeat(64);
const TERMINAL_EVENTS = new Set(["task.stop", "task.fail"]);

export class TraceVerificationError extends Error {}
export class ToolAuthorizationError extends Error {}

export function canonicalJson(value: unknown): string {
  const encoded = canonicalize(value);
  if (encoded === undefined) {
    throw new Error("value is not representable in RFC 8785 JCS");
  }
  return encoded;
}

export function sha256Hex(value: unknown): string {
  const text = typeof value === "string" ? value : canonicalJson(value);
  return createHash("sha256").update(text, "utf8").digest("hex");
}

export function computeEventHash(event: EventRecord | JsonObject): string {
  const material = { ...(event as JsonObject) };
  delete material.event_hash;
  return sha256Hex(material);
}

export type ToolPermit = {
  toolName: string;
  requestEventId: string;
  permitEventId: string;
};

export class Monitor {
  private sequence = 0;
  readonly events: EventRecord[] = [];
  readonly runId: string;
  readonly anchors: EventRecord["anchors"];
  private readonly allowedTools: Set<string>;

  constructor(
    readonly agentId: string,
    readonly taskId: string,
    readonly objective: string,
    readonly policy: JsonObject,
    readonly role = "actor",
    allowedTools: Iterable<string> = [],
    runId?: string,
    readonly rootAgentId = agentId,
    readonly parentAgentId: string | null = null,
  ) {
    if (!agentId || !taskId || !objective || !role) {
      throw new Error("agentId, taskId, objective, and role are required");
    }
    this.runId = runId ?? randomUUID();
    this.allowedTools = new Set(allowedTools);
    this.anchors = {
      role_hash: sha256Hex(role),
      objective_hash: sha256Hex(objective),
      policy_hash: sha256Hex(policy),
    };
  }

  private emit(eventType: string, payload: JsonObject = {}): EventRecord {
    if (
      this.events.length > 0 &&
      TERMINAL_EVENTS.has(this.events[this.events.length - 1].event_type)
    ) {
      throw new Error("cannot emit after a terminal event");
    }
    this.sequence += 1;
    const event = {
      event_id: randomUUID(),
      event_type: eventType,
      timestamp: new Date().toISOString(),
      agent_id: this.agentId,
      task_id: this.taskId,
      run_id: this.runId,
      sequence: this.sequence,
      anchors: { ...this.anchors },
      prev_event_hash:
        this.events.length > 0
          ? this.events[this.events.length - 1].event_hash
          : ZERO_HASH,
      payload: structuredClone(payload),
      event_hash: "",
    };
    event.event_hash = computeEventHash(event);
    this.events.push(event);
    return structuredClone(event);
  }

  taskStart(metadata: JsonObject = {}): EventRecord {
    if (this.events.length !== 0) {
      throw new Error("task.start must be the first event");
    }
    return this.emit("task.start", {
      schema_version: "1.1.0",
      root_agent_id: this.rootAgentId,
      parent_agent_id: this.parentAgentId,
      role: this.role,
      objective_digest: this.anchors.objective_hash,
      policy_digest: this.anchors.policy_hash,
      ...metadata,
    });
  }

  snapshotState(
    inputs: JsonObject,
    outputs: JsonObject,
    confidence: JsonObject = {},
    assumptions: string[] = [],
    evidenceRefs: string[] = [],
  ): EventRecord {
    const nextSequence = this.sequence + 1;
    return this.emit("state.snapshot", {
      schema_version: "1.1.0",
      identity: {
        agent_id: this.agentId,
        root_agent_id: this.rootAgentId,
        parent_agent_id: this.parentAgentId,
        role: this.role,
        role_hash: this.anchors.role_hash,
      },
      intent: {
        objective_hash: this.anchors.objective_hash,
        policy_hash: this.anchors.policy_hash,
      },
      trace: {
        run_id: this.runId,
        task_id: this.taskId,
        step_id: `step-${nextSequence}`,
        sequence: nextSequence,
        parent_event_hash:
          this.events.length > 0
            ? this.events[this.events.length - 1].event_hash
            : ZERO_HASH,
      },
      inputs,
      outputs,
      tools: [],
      confidence,
      policy: this.policy,
      assumptions,
      evidence_refs: evidenceRefs,
    });
  }

  requestTool(toolName: string, args: JsonObject): ToolPermit {
    const argumentsDigest = sha256Hex(args);
    const request = this.emit("tool.requested", {
      tool_name: toolName,
      arguments_digest: argumentsDigest,
    });
    const decision = {
      tool_name: toolName,
      request_event_id: request.event_id,
      arguments_digest: argumentsDigest,
      policy_hash: this.anchors.policy_hash,
    };
    if (!this.allowedTools.has(toolName)) {
      this.emit("tool.denied", decision);
      throw new ToolAuthorizationError(`tool denied by policy: ${toolName}`);
    }
    const permit = this.emit("tool.authorized", decision);
    return {
      toolName,
      requestEventId: request.event_id,
      permitEventId: permit.event_id,
    };
  }

  recordToolResult(
    permit: ToolPermit,
    result: unknown,
    failed = false,
  ): EventRecord {
    return this.emit(failed ? "tool.failed" : "tool.executed", {
      tool_name: permit.toolName,
      request_event_id: permit.requestEventId,
      permit_event_id: permit.permitEventId,
      result_digest: sha256Hex(result),
    });
  }

  recordApprovalDecision(
    actorId: string,
    criticId: string,
    judgeId: string,
    proposal: unknown,
    critique: unknown,
    decision: "approved" | "rejected",
    evidenceRefs: string[],
  ): EventRecord {
    if (
      decision === "approved" &&
      new Set([actorId, criticId, judgeId]).size !== 3
    ) {
      throw new Error(
        "positive approval requires pairwise-distinct Actor/Critic/Judge",
      );
    }
    if (decision === "approved" && evidenceRefs.length === 0) {
      throw new Error("positive approval requires evidence");
    }
    return this.emit("approval.decided", {
      actor_id: actorId,
      critic_id: criticId,
      judge_id: judgeId,
      proposal_hash: sha256Hex(proposal),
      critique_hash: sha256Hex(critique),
      decision,
      evidence_refs: evidenceRefs,
    });
  }

  taskStop(summary?: string): EventRecord {
    return this.emit("task.stop", summary ? { summary } : {});
  }

  taskFail(
    failureClass: string,
    failureSummary: string,
    impact: string,
  ): EventRecord {
    return this.emit("task.fail", {
      failure_class: failureClass,
      failure_summary: failureSummary,
      impact,
    });
  }
}

export function verifyTrace(
  events: EventRecord[],
  allowedTools?: Iterable<string>,
): void {
  if (events.length === 0) {
    throw new TraceVerificationError("empty trace");
  }
  if (events[0].event_type !== "task.start") {
    throw new TraceVerificationError("trace must begin with task.start");
  }

  const expectedAnchors = canonicalJson(events[0].anchors);
  const allowed = allowedTools ? new Set(allowedTools) : undefined;
  const permits = new Map<
    string,
    { requestEventId: string; toolName: string }
  >();
  const denied = new Set<string>();
  let previousHash = ZERO_HASH;
  let terminalSeen = false;

  events.forEach((event, offset) => {
    const sequence = offset + 1;
    if (terminalSeen) {
      throw new TraceVerificationError("event appears after terminal event");
    }
    if (event.sequence !== sequence) {
      throw new TraceVerificationError("non-contiguous event sequence");
    }
    if (event.prev_event_hash !== previousHash) {
      throw new TraceVerificationError("broken previous-event hash link");
    }
    if (event.event_hash !== computeEventHash(event)) {
      throw new TraceVerificationError("event hash mismatch");
    }
    if (canonicalJson(event.anchors) !== expectedAnchors) {
      throw new TraceVerificationError("immutable anchor drift");
    }

    const payload = event.payload;
    if (event.event_type === "tool.denied") {
      const requestId = payload.request_event_id;
      if (typeof requestId === "string") denied.add(requestId);
    } else if (event.event_type === "tool.authorized") {
      const requestId = payload.request_event_id;
      const toolName = payload.tool_name;
      if (typeof requestId !== "string" || typeof toolName !== "string") {
        throw new TraceVerificationError("malformed tool authorization");
      }
      if (denied.has(requestId)) {
        throw new TraceVerificationError("request both denied and authorized");
      }
      if (allowed && !allowed.has(toolName)) {
        throw new TraceVerificationError(
          "authorization violates verifier policy",
        );
      }
      permits.set(event.event_id, {
        requestEventId: requestId,
        toolName,
      });
    } else if (
      event.event_type === "tool.executed" ||
      event.event_type === "tool.failed"
    ) {
      const permitId = payload.permit_event_id;
      const requestId = payload.request_event_id;
      const toolName = payload.tool_name;
      if (typeof permitId !== "string") {
        throw new TraceVerificationError("tool execution without valid permit");
      }
      const permit = permits.get(permitId);
      if (!permit) {
        throw new TraceVerificationError("tool execution without valid permit");
      }
      if (
        permit.requestEventId !== requestId ||
        permit.toolName !== toolName
      ) {
        throw new TraceVerificationError(
          "tool execution does not match permit",
        );
      }
      if (typeof requestId === "string" && denied.has(requestId)) {
        throw new TraceVerificationError("denied tool request was executed");
      }
      if (
        allowed &&
        typeof toolName === "string" &&
        !allowed.has(toolName)
      ) {
        throw new TraceVerificationError(
          "executed tool violates verifier policy",
        );
      }
    } else if (
      event.event_type === "approval.decided" &&
      payload.decision === "approved"
    ) {
      const ids = new Set([
        payload.actor_id,
        payload.critic_id,
        payload.judge_id,
      ]);
      if (ids.has(undefined) || ids.size !== 3) {
        throw new TraceVerificationError(
          "self-approval or missing SDV identity",
        );
      }
      if (!payload.proposal_hash || !payload.critique_hash) {
        throw new TraceVerificationError(
          "approval missing proposal/critique digest",
        );
      }
      if (
        !Array.isArray(payload.evidence_refs) ||
        payload.evidence_refs.length === 0
      ) {
        throw new TraceVerificationError("approval missing evidence");
      }
    }

    if (TERMINAL_EVENTS.has(event.event_type)) {
      terminalSeen = true;
    }
    previousHash = event.event_hash;
  });

  if (!TERMINAL_EVENTS.has(events[events.length - 1].event_type)) {
    throw new TraceVerificationError("trace is not terminal");
  }
}
